pkg/reconciler/pipeline/validate.go

```diff
@@ -472,3 +476,37 @@ func toVerticesMappedByFrom(edges []dfv1.Edge, verticesByName map[string]*dfv1.A
 	}
 	return mappedEdges, nil
 }
+
+func findForests(vtx dfv1.AbstractVertex, pl *dfv1.Pipeline) bool {
+
+	visited := map[string]struct{}{}
+	findForestHelper(vtx.Name, visited, pl)
```

Comment:
nit - naming: we have findForest**s** and findForestHelper, do we need the extra s? Also consider naming it `isAForest`